src/executables/loadjava.cpp

```diff
@@ -0,0 +1,239 @@
+
+#include "cubrid_getopt.h"
+#include "error_code.h"
+#include "message_catalog.h"
+#include "utility.h"
+#include "databases_file.h"
+#if defined(WINDOWS)
+#include "porting.h"
+#endif /* WINDOWS */
+
+namespace fs = std::filesystem;
+using namespace std::regex_constants;
+
+#define JAVA_DIR                "java"
+#if defined(WINDOWS)
+#define SEPERATOR               "\\"
+#else /* ! WINDOWS */
+#define SEPERATOR               "/"
+#endif /* !WINDOWS */
+
+static const std::string JAVA_PACKAGE_PATTERN = "^([a-z_]{1}[a-z0-9_]*(\\.[a-z_]{1}[a-z0-9_]*)*)$";
+static const std::string SEPARATOR_STRING (SEPERATOR);
+
+static char *Program_name = NULL;
+static char *Dbname = NULL;
+std::string Src_class;
+std::regex Java_package_reg (JAVA_PACKAGE_PATTERN, ECMAScript | icase | optimize);
+
+static int Force_overwrite = false;
+static std::string package_path;
+
+static void
+usage (void)
+{
+  fprintf (stderr, "Usage: loadjava [OPTION] database-name java-class-file\n");
+  fprintf (stderr, "Options:\n-y\t%s\n-p\t%s\n",
+	   msgcat_message (MSGCAT_CATALOG_UTILS, MSGCAT_UTIL_SET_LOADJAVA, LOADJAVA_ARG_FORCE_OVERWRITE_HELP),
+	   msgcat_message (MSGCAT_CATALOG_UTILS, MSGCAT_UTIL_SET_LOADJAVA, LOADJAVA_ARG_FORCE_PACKAGE_HELP));
+}
+
+static int
+parse_argument (int argc, char *argv[])
+{
+  struct option loadjava_option[] =
+  {
+    {"overwrite", 0, 0, 'y'},
+    {"package", 0, 0, 'p'},
+    {0, 0, 0, 0}
+  };
+
+  while (1)
+    {
+      int option_index = 0;
+      int option_key = getopt_long (argc, argv, "y:p:h", loadjava_option, &option_index);
+      if (option_key == -1)
+	{
+	  break;
+	}
+
+      switch (option_key)
+	{
+	case 'y':
+	  Force_overwrite = true;
+	  break;
+	case 'p':
+	{
+	  // check valid package name
+	  std::string package_name (optarg);
+	  if (!package_name.empty())
+	    {
+	      bool is_matched = std::regex_search (package_name, Java_package_reg);
+	      if (!is_matched)
+		{
+		  fprintf (stderr, "invalid java package name\n");
+		  return ER_FAILED;
+		}
+	      // replace all for package name's dot to SEPARATER
+	      // e.g. org.cubrid.abc => /org/cubrid/abc/
+	      package_path = SEPERATOR;
+	      package_path += std::regex_replace (package_name, std::regex ("\\."), SEPARATOR_STRING);
+	      package_path += SEPERATOR;
+	    }
+	}
+	break;
+	case 'h':
+	/* fall through */
+	default:
+	  usage ();
+	  return ER_FAILED;
+	}
+    }
+
+  if (optind + 1 < argc)
+    {
+      Dbname = argv[optind];
+      Src_class = argv[optind + 1];
+    }
+  else
+    {
+      usage ();
+      return ER_FAILED;
+    }
+
+  Program_name = argv[0];
+
+  return NO_ERROR;
+}
+
+static int
+create_package_directories (const std::string &java_dir_path)
+{
+  try
+    {
+      if (fs::exists (java_dir_path) == false)
+	{
+	  fs::create_directories (java_dir_path);
+	  fs::permissions (java_dir_path,
+			   fs::perms::owner_all | fs::perms::group_read | fs::perms::others_read,
+			   fs::perm_options::add);	// mkdir (java_dir_path, 0744)
```

Comment:
Is there a reason for 744 instead of giving 755?
It seems to be the same as 700 if the execute permission is missing from the directory.

Reply:
I just keep it as it is. the classes in the path are not executable, IMO the permission seems not required.

Reply:
It is trivial.

Usually, it is natural to check the umask and grant permission.

The owner has all the rights, so no problem.

However, neither group users nor other users can enter the directory because they do not have execute permission on the directory.
So, I know that directories are given read permission and execute permission together or not together.